### DISCUSSION TITLE
How to access view-model properties when using shapes?

### DISCUSSION THREAD

**Discussion by MikeAlhayek** [OrchardCMS/OrchardCore#8219] [Q&A]
I am trying to create a simple CRUD function and render the Shapes using razor views. I am trying to follow [OrchardSkills.OrchardCore.CRUD](https://github.com/OrchardSkills/OrchardSkills.OrchardCore.CRUD/blob/master/OrchardCore.CRUD/Views/Home/Index.cshtml) but instead of using `ContentItem` as a model, I am using my own model.

I created a driver that would initialize different view-model based on the display-type.

```
public override Task<IDisplayResult> DisplayAsync(Biometric model, BuildDisplayContext context)
{
    if(context.DisplayType == "Summary")
    {
        return Task.FromResult<IDisplayResult>(Initialize<DisplayBiometricSummaryViewModel>(viewModel =>
        {
            viewModel.Id = biometric.Id;
            viewModel.IsActive = biometric.IsActive;
            viewModel.MeasuredAt = biometric.MeasuredAt;
            viewModel.Weight = biometric.Weight;
        }));
    }

    return Task.FromResult<IDisplayResult>(Initialize<DisplayBiometricViewModel>(viewModel =>
    {
        viewModel.Id = biometric.Id;
        viewModel.IsActive = biometric.IsActive;
        viewModel.MeasuredAt = biometric.MeasuredAt;
        viewModel.Weight = biometric.Weight;
    }));
}
```

Here is my ListBiometricViewModel

```
    public class ListBiometricViewModel
    {
        public int? Page { get; set; }

        [BindNever]
        public List<IShape> Records { get; set; }

        [BindNever]
        public dynamic Pager { get; set; }

        public ListBiometricViewModel()
        {
            Records = new List<IShape>();
        }
    }
```

In my `Index.cshtml` view, I am rending the `ListBiometricViewModel` like this

```
@model ListBiometricViewModel

 <table class="table table-sm table-striped">
      <tbody>
          @foreach (var record in Model.Records)
          {
              @await DisplayAsync(record)
          }
      </tbody>
  </table>

```

Then I created a view called `Biometric.Summary.cshtml`  with the following code, but 

```
    <tr>
        <td></td>
        <td>@Model.Weight</td>
        <td>@Model.MeasuredAt</td>
        <td>@Model.IsActive</td>
    </tr>
```

I don't think I am accessing the model properties correctly, wish is why I am seeing no output. When adding a break point into the views, I do see them being called, but not sure how to get the data.

How can I access the properties of my shape?

**Comment by jtkech** [comment]
You don't need to check the display type in your driver, and you only need one `DisplayBiometricViewModel`

Hmm, then try to use in your `ListBiometricViewModel`

        public List<Biometric> Records { get; set; }

Then, as i remember, you will need to build a shape for each one and render it, see in the `ListPart.cshtml` of the `OC.Lists` module how it is done.

    @foreach (var contentItem in Model.ContentItems)
    {
        var contentItemSummary = await ContentItemDisplayManager.BuildDisplayAsync(contentItem, Model.Context.Updater, "Summary", Model.Context.GroupId);

        <li class="list-group-item">
            @await DisplayAsync(contentItemSummary)
        </li>
    }

Then normally in your `Biometric` view you can dynamically access the properties, or try to use

    @model DisplayBiometricViewModel

**Reply by MikeAlhayek** [reply]
@jtkech  I am trying to figure out how OC.Lists constructs the `ListPartViewModel` class before it is passed to the `ListPart.cshtml` view. I see the [ListPartDisplayDriver](https://github.com/OrchardCMS/OrchardCore/blob/8280e2ba8caaadfbf6f72e4ee998d8cda9bb441e/src/OrchardCore.Modules/OrchardCore.Lists/Drivers/ListPartDisplayDriver.cs#L34-L80) drivers which creates that `ListPartViewModel`. But, at what point the `ListPartDisplayDriver` is called before it is passed to the `ListPart.cshtml`?

**Reply by MikeAlhayek** [reply]
@jtkech I am still unsure how `OC.Lists` constructs the `ListPartViewModel`. Additionally, I made progress explained in #8210, I would appreciate it, if you can help me with the questions I posted in the [discussion](https://github.com/OrchardCMS/OrchardCore/discussions/8210#discussioncomment-260588)

**Reply by jtkech** [reply]
@CrestApps 

> But, at what point the ListPartDisplayDriver is called before it is passed to the ListPart.cshtml?

When the display manager builds the shape of a given content item, all part / field drivers are coordinated and collaborate to the shape composition, then the `ListPartDisplayDriver` combine shapes whose one does

                        model.ContentItems = (await _containerService.QueryContainedItemsAsync(
                            listPart.ContentItem.ContentItemId,
                            settings.EnableOrdering,
                            pager,
                            containedItemOptions)).ToArray();

Here the content items list is not a list of shapes, there are content items, and this is in the `ListPart.cshtml` that there are built as shapes and then rendered.

    @foreach (var contentItem in Model.ContentItems)
    {
        var contentItemSummary = await ContentItemDisplayManager.BuildDisplayAsync(contentItem, Model.Context.Updater, "Summary", Model.Context.GroupId);

        <li class="list-group-item">
            @await DisplayAsync(contentItemSummary)
        </li>
    }

---

Then about your list

        public List<IShape> Records { get; set; }

It depends how you built them, where do you build them and how? are your drivers called?

For a regular content item, it is done in the `ItemController` that uses a content item display manager to build / compose the full shape by calling part / field drivers (that build their own child shapes) / handlers. Then as said above, a shape view can also build itself other shapes e.g. to render a list of items.

To not have to build a shape for a given type, you can also derive it from `ShapeViewModel` (that is an `IShape`), and then you can do a shape rendering on it directly without having to build a shape.
